### PR TITLE
Add unitest

### PR DESCRIPTION
**Pull request by passer-byzhang**
unitest for `depositeToken`,`claimToken` and `withdrawToken`

### CODE REVIEW

File: contracts/tests/MockERC20.sol [hunk 1]
```diff
@@ -0,0 +1,12 @@
...
+import "@openzeppelin/contracts/token/ERC20/ERC20.sol";
+contract MockERC20 is ERC20 {
+    constructor(string memory name, string memory symbol) ERC20(name, symbol) {
+        _mint(msg.sender, 1000000000000000000000000);
+    }
+
+    function mint(address to, uint256 amount)  external {
+        _mint(to, amount);
+    }
+}
```
Comment (mialbu) [inline contracts/tests/MockERC20.sol]: empty line

File: contracts/tests/SigUtils.sol [hunk 2]
```diff
@@ -0,0 +1,53 @@
...
+    uint256 internal user0PrivateKey = 0xac0974bec39a17e36ba4a6b4d238ff944bacb478cbed5efcae784d7bf4f2ff80;
+    uint256 internal user1PrivateKey = 0x59c6995e998f97a5a0044966f0945389dc9e86dae88c7a8412f4603b6b78690d;
+    uint256 internal user2PrivateKey = 0x5de4111afa1a4b94908f83103eb1f1706367c2e68ca870fc3fb9a804cdab365a;
+    uint256 internal user3PrivateKey = 0x7c852118294e51e653712a81e05800f419141751be58f605c371e15141b007a6;
+    uint256 internal user4PrivateKey = 0x47e179ec197488593b187f80a00eb0da91f1b9d0b13f8733639f19c30a34926a;
+    uint256 internal user5PrivateKey = 0x8b3a350cf5c34c9194ca85829a2df0ec3153be0318b5e2d3348e872092edffba;
+    uint256 internal user6PrivateKey = 0x92db14e403b83dfe3df233f83dfa3a0d7096f21ca9b0d6d6b8d88b2b4ec1564e;
+
+
+
```
Comment (mialbu) [inline contracts/tests/SigUtils.sol]: format, 1 empty line is sufficient.

File: contracts/tests/SigUtils.sol [hunk 3]
```diff
@@ -0,0 +1,53 @@
+// SPDX-License-Identifier: MIT
+pragma solidity ^0.8.24;
+import "../library/BridgeLib.sol";
```
Comment (mialbu) [inline contracts/tests/SigUtils.sol:3]: format, empty lines between pragma/import/contract.

File: contracts/tests/SigUtils.sol [hunk 4]
```diff
@@ -0,0 +1,53 @@
...
+        internal
+        pure
+        returns (bytes32)
+    {
+        return
+            keccak256(
+                abi.encodePacked(
+                    _depositData.to,
+                    _depositData.amount,
+                    _depositData.nonce
```
Comment (mialbu) [inline contracts/tests/SigUtils.sol]: This order looks wrong. The concatenation is in the order `nonce`, `to`, `amount`. The concatenation here looks like the reverse of the previous concatenation order. Is there a reason it was reversed?

File: contracts/tests/TestBridge.sol [hunk 5]
```diff
@@ -0,0 +1,64 @@
+// SPDX-License-Identifier: MIT
+pragma solidity ^0.8.0;
+import "../bridge/BridgeImpl.sol";
```
Comment (mialbu) [inline contracts/tests/TestBridge.sol:3]: format

File: .states/vm-cancun/state.json [hunk 6]
```diff
@@ -0,0 +1,61 @@
...
+	],
+	"latestBlockNumber": "0x1"
+}
```
Comment (mialbu) [inline .states/vm-cancun/state.json]: Please add an empty lines in every file.

File: contracts/tests/SigUtils.sol [hunk 7]
```diff
@@ -0,0 +1,53 @@
...
+    {
+        return keccak256(
+            abi.encodePacked(
+                "\x19Ethereum Signed Message:\n32",
+                keccak256(abi.encodePacked(_hash))
+            )
+        );
+    }
+
+}
```
Comment (mialbu) [inline contracts/tests/SigUtils.sol]: empty line

File: contracts/tests/TestBridge.sol [hunk 8]
```diff
@@ -0,0 +1,64 @@
...
+
+    function getTokenbridge(
+        address neoXToken
+    ) public view returns (StorageTypes.TokenBridge memory tokenBridgeBefore) {
+        return tokenBridges[neoXToken];
+    }
+
+
+
+
```
Comment (mialbu) [inline contracts/tests/TestBridge.sol]: format. ok I stop commenting this. Please just make sure to remove blocks of multiple unnecessary empty lines, and make sure each file has an empty line at the end of the file.

File: test/Bridge.t.sol [hunk 9]
```diff
@@ -0,0 +1,361 @@
...
+
+    address neoXToken = address(0x6789);
+    address neoN3Token = address(0x7892);
+    StorageTypes.TokenConfig validConfig;
+
+    // set _management
+    BridgeManagementImpl bridgeManagementImpl;
+    SigUtils sigUtils;
+    address public owner = 0xBcd4042DE499D14e55001CcbB24a551F3b954096;
+    address public user = 0xf39Fd6e51aad88F6F4ce6aB8827279cffFb92266;
```
Comment (mialbu) [inline test/Bridge.t.sol]: The `user` has the role of the funder, please rename it to `funder` and use different address than relayer.

File: test/Bridge.t.sol [hunk 10]
```diff
@@ -0,0 +1,361 @@
...
+            minAmount: 100,
+            maxAmount: 1000,
+            maxDeposits: 10,
+            executionType: StorageTypes.ExecutionType.NEO
+        });
+    }
+
+    function testRegisterToken() public {
+        vm.prank(governor);
+        // console.log("msg.sender: %s, token: %s",msg.sender,neoXToken);
```
Comment (mialbu) [inline test/Bridge.t.sol]: remove

File: contracts/tests/TestBridge.sol
Comment (mialbu) [inline contracts/tests/TestBridge.sol:1]: Does Foundry not support reading public variables, or why do we need these wrapped functions?

For the boolean `isRegisteredToken` I get the point of having it, since it has additional logic to just reading. However, I don't see the point for the other getters, e.g., `getNeoN3Token`.

File: contracts/tests/TestBridge.sol [hunk 11]
```diff
@@ -0,0 +1,64 @@
...
+        return bridgePaused;
+    }
+
+
+    function computeTokenRoot(
+        bytes32 _previousRoot,
+        address _neoN3Token,
+        address _neoXToken,
+        BridgeLib.DepositData[] calldata _deposits
+    ) public returns (bytes32) {
```
Comment (mialbu) [inline contracts/tests/TestBridge.sol]:
```suggestion
    ) public pure returns (bytes32) {
```

File: test/Bridge.t.sol [hunk 12]
```diff
@@ -0,0 +1,361 @@
...
+        vm.prank(governor);
+        bridgeImpl.registerToken(neoXToken, validConfig);
+        vm.expectRevert(
+            abi.encodeWithSelector(
+                BridgeStorage.TokenBridgeAlreadyRegistered.selector,
+                neoXToken
+            )
+        );
+        vm.prank(governor);
+        bridgeImpl.registerToken(neoXToken, validConfig); // Should fail
```
Comment (mialbu) [inline test/Bridge.t.sol]:
```suggestion
        bridgeImpl.registerToken(neoXToken, validConfig);
```

File: test/Bridge.t.sol [hunk 13]
```diff
@@ -0,0 +1,361 @@
...
+    }
+
+    function test_RegisterTokenByNonGovernorr() public {
```
Comment (mialbu) [inline test/Bridge.t.sol]:
```suggestion
    function test_RegisterTokenByNonGovernor() public {
```

File: test/Bridge.t.sol [hunk 14]
```diff
@@ -0,0 +1,361 @@
...
+        bridgeImpl.pauseTokenBridge(neoXToken);
+
+        // Verify the token bridge is paused
+        StorageTypes.TokenBridge memory tokenBridgeBefore = bridgeImpl
+            .getTokenbridge(neoXToken);
+        assertTrue(tokenBridgeBefore.paused);
+
+        // Unregister the token
+        // vm.expectEmit(true, true, true, true);
+        // emit bridgeImpl.TokenUnregister(neoXToken,validConfig);
```
Comment (mialbu) [inline test/Bridge.t.sol]: Can you add an actual check if the event is emitted as expected?

File: test/Bridge.t.sol [hunk 15]
```diff
@@ -0,0 +1,361 @@
...
+        vm.prank(governor);
+        bridgeImpl.unregisterToken(neoXToken);
+
+        // // Verify that the token bridge is unregistered
```
Comment (mialbu) [inline test/Bridge.t.sol]:
```suggestion
        // Verify that the token bridge is unregistered
```

File: test/Bridge.t.sol [hunk 16]
```diff
@@ -0,0 +1,361 @@
...
+        // Verify the token bridge is paused
+        StorageTypes.TokenBridge memory tokenBridgeBefore = bridgeImpl
+            .getTokenbridge(neoXToken);
+        assertTrue(tokenBridgeBefore.paused);
+        vm.prank(governor);
+        bridgeImpl.unregisterToken(neoXToken);
+
+        vm.prank(governor);
+        vm.expectRevert(
+            abi.encodeWithSignature("TokenBridgeUnpaused(address)", neoXToken)
```
Comment (mialbu) [inline test/Bridge.t.sol:244]: 👍🏼 good test. Got me thinking, maybe we should use a formal check here if a token is registered instead of reverting with `TokenBridgeUnpaused`. Reverting with `TokenBridgeNotRegistered` is better suited here.

File: test/Bridge.t.sol [hunk 17]
```diff
@@ -0,0 +1,361 @@
...
+
+        // Attempt to pause an unregistered token bridge
+        vm.prank(securityGuard);
+        vm.expectRevert(
+            abi.encodeWithSignature(
+                "TokenBridgeNotRegistered(address)",
+                unregisteredToken
+            )
+        );
+        // vm.expectRevert(abi.encodeWithSelector(BridgeStorage.TokenBridgeNotRegistered.selector,unregisteredToken));
```
Comment (mialbu) [inline test/Bridge.t.sol]:
```suggestion
```

File: test/Bridge.t.sol [hunk 18]
```diff
@@ -0,0 +1,361 @@
+// SPDX-License-Identifier: MIT
+pragma solidity ^0.8.0;
+import "../lib/forge-std/src/Test.sol";
+import {TestBridge} from "../contracts/tests/TestBridge.sol";
+import "../contracts/bridge/BridgeStorage.sol";
+import "../contracts/management/BridgeManagementImpl.sol";
+import "../contracts/tests/SigUtils.sol";
+
+contract BridgeImplTest is Test, SigUtils {
+    // constructor(address _management) BridgeImpl(_management) {}
```
Comment (mialbu) [inline test/Bridge.t.sol]: Please move this to where the `TestBridge` is instantiated, i.e., line 55.

File: test/BridgeManagement.t.sol [hunk 19]
```diff
@@ -0,0 +1,81 @@
...
+import "../lib/forge-std/src/Test.sol";
+import "../contracts/management/BridgeManagementImpl.sol";
+import "../contracts/tests/SigUtils.sol";
+import "../contracts/library/BridgeLib.sol";
+
+contract BridgeManagementImplTest is Test, SigUtils {
+    BridgeManagementImpl bridgeManagementImpl;
+    SigUtils sigUtils;
+    address public owner = 0xBcd4042DE499D14e55001CcbB24a551F3b954096;
+    address public user = 0xf39Fd6e51aad88F6F4ce6aB8827279cffFb92266;
```
Comment (mialbu) [inline test/BridgeManagement.t.sol]: rename to `funder` and use different address than relayer.

File: test/Bridge.t.sol [hunk 20]
```diff
@@ -0,0 +1,361 @@
...
+    StorageTypes.TokenConfig validConfig;
+
+    // set _management
+    BridgeManagementImpl bridgeManagementImpl;
+    SigUtils sigUtils;
+    address public owner = 0xBcd4042DE499D14e55001CcbB24a551F3b954096;
+    address public user = 0xf39Fd6e51aad88F6F4ce6aB8827279cffFb92266;
+    address public relayer = 0xf39Fd6e51aad88F6F4ce6aB8827279cffFb92266;
+    uint8 public validatorThreshold = 5;
+    uint256[] public validatorsKeys;
```
Comment (mialbu) [inline test/Bridge.t.sol]: Not needed in this test, please remove it.

File: test/BridgeManagement.t.sol [hunk 21]
```diff
@@ -0,0 +1,81 @@
...
+        BridgeLib.Signature[] memory _signatures = new BridgeLib.Signature[](5);
+        bytes32 ethHash = getSignedHash(_depositRoot);
+        for (uint i = 0; i < 5; i++) {
+            (uint8 v, bytes32 r, bytes32 s) = vm.sign(
+                validatorsKeys[i],
+                ethHash
+            );
+            address x = ecrecover(ethHash, v, r, s);
+            _signatures[i] = BridgeLib.Signature(v, r, s);
+            assertEq(x, validatorsAddresses[i]);
```
Comment (mialbu) [inline test/BridgeManagement.t.sol]:
```suggestion
            address recoveredAddr = ecrecover(ethHash, v, r, s);
            _signatures[i] = BridgeLib.Signature(v, r, s);
            assertEq(recoveredAddr, validatorsAddresses[i]);
```

File: test/FungibleToken.t.sol [hunk 22]
```diff
@@ -0,0 +1,764 @@
...
+pragma solidity ^0.8.0;
+import "../lib/forge-std/src/Test.sol";
+import {TestBridge} from "../contracts/tests/TestBridge.sol";
+import {BridgeStorage, BridgeLib, GasBridgeLib, StorageTypes, TokenBridgeLib} from "../contracts/bridge/BridgeStorage.sol";
+import "../contracts/management/BridgeManagementImpl.sol";
+import "../contracts/tests/SigUtils.sol";
+import "../contracts/tests/MockERC20.sol";
+
+contract TestFungibleToken is Test, SigUtils {
+    // constructor(address _management) BridgeImpl(_management) {}
```
Comment (mialbu) [inline test/FungibleToken.t.sol]: move to instantiation.

File: test/FungibleToken.t.sol [hunk 23]
```diff
@@ -0,0 +1,764 @@
...
+    address neoN3TokenA = address(0x7892);
+    address neoN3TokenB = address(0x7893);
+    StorageTypes.TokenConfig validConfigA;
+    StorageTypes.TokenConfig validConfigB;
+
+    // set _management
+    BridgeManagementImpl bridgeManagementImpl;
+    SigUtils sigUtils;
+    address public owner = 0xBcd4042DE499D14e55001CcbB24a551F3b954096;
+    address public user = 0xf39Fd6e51aad88F6F4ce6aB8827279cffFb92266;
```
Comment (mialbu) [inline test/FungibleToken.t.sol]: rename to `funder` and use different address than relayer. Also, please use different addresses for transfer recipients, i.e., addresses that don't have any role in the bridge management.

File: test/FungibleToken.t.sol [hunk 24]
```diff
@@ -0,0 +1,764 @@
...
+
+        assertEq(MockERC20(neoXTokenB).balanceOf(user), balance - 300 ether);
+        assertEq(MockERC20(neoXTokenB).balanceOf(owner), balance - 700 ether);
+        //@TODO
+        //check withdrwal state and missing max/min deposit amount check in depositToken?
+    }
+
+    function getSignatures(
+        bytes32 _depositRoot
+    ) public view returns (BridgeLib.Signature[] memory) {
```
Comment (mialbu) [inline test/FungibleToken.t.sol]: This always takes the signatures of the first 5 validators. I think we should at least have 1 or 2 tests where validators 6 and 7 are also involved. Maybe you could add an int array of size 5 as parameter (for indexes), so that you can specify which validators should sign the root.

File: test/FungibleToken.t.sol [hunk 25]
```diff
@@ -0,0 +1,764 @@
...
+        BridgeLib.Signature[] memory signatures = getSignatures(
+            tokenDepositRoot
+        );
+        vm.prank(relayer);
+        bridgeImpl.depositToken(
+            neoXTokenA,
+            tokenDepositRoot,
+            signatures,
+            depositData
+        );
```
Comment (mialbu) [inline test/FungibleToken.t.sol:277]: Can you add a check that tests the `TokenClaimable` event here?

File: test/FungibleToken.t.sol [hunk 26]
```diff
@@ -0,0 +1,764 @@
...
+
+    function testDepositTokenA() public {
+        MockERC20(neoXTokenA).mint(address(bridgeImpl), 100 ether);
+        vm.prank(governor);
+        bridgeImpl.registerToken(neoXTokenA, validConfigA);
+        BridgeLib.DepositData[]
+            memory depositData = new BridgeLib.DepositData[](2);
+        BridgeLib.DepositData memory d0 = BridgeLib.DepositData({
+            to: payable(user),
+            amount: 100,
```
Comment (mialbu) [inline test/FungibleToken.t.sol]: Maybe use different value than `100` here - differentiating a bit more with the `100 ether` mint in the beginning of the test. I overlooked the `ether` and expected a different test case. Or leave it like it is and just add a comment at the beginning of the test stating the obvious. 🙂

File: test/FungibleToken.t.sol [hunk 27]
```diff
@@ -0,0 +1,764 @@
...
+        );
+        bridgeImpl.claimToken(neoXTokenB, 1);
+    }
+
+    function testWithdrawTokenA() public {
+        console.log(
+            "neoxtoken address: %s,neo address : %s, nonce = 0",
+            neoXTokenA,
+            neoN3TokenA
+        );
```
Comment (mialbu) [inline test/FungibleToken.t.sol]:
```suggestion
```

File: test/FungibleToken.t.sol [hunk 28]
```diff
@@ -0,0 +1,764 @@
...
+        console.log("user: %s,amount: %d", owner, 400);
+        bridgeImpl.withdrawToken{value: validConfigA.fee}(
+            neoXTokenA,
+            owner,
+            400
+        );
+        assertEq(MockERC20(neoXTokenA).balanceOf(user), balance - 300);
+        assertEq(MockERC20(neoXTokenA).balanceOf(owner), balance - 700);
+        //@TODO
+        //check withdrwal state and missing max/min deposit amount check in depositToken?
```
Comment (mialbu) [inline test/FungibleToken.t.sol]: Yes, such tests would be great.

File: test/FungibleToken.t.sol [hunk 29]
```diff
@@ -0,0 +1,764 @@
...
+        );
+        bridgeImpl.depositToken(
+            neoXTokenA,
+            tokenDepositRoot,
+            signatures,
+            depositData
+        );
+    }
+
+    // todo deposittoken到底需不需要relayer
```
Comment (mialbu) [inline test/FungibleToken.t.sol]: I don't understand the todo 🥲 please commit in english 😅

File: test/FungibleToken.t.sol [hunk 30]
```diff
@@ -0,0 +1,764 @@
...
+            bridgeImpl.getTokenDepositState(neoXTokenA).root,
+            neoN3TokenA,
+            neoXTokenA,
+            depositData
+        );
+        BridgeLib.Signature[] memory signatures = getSignatures(
+            tokenDepositRoot
+        );
+        vm.prank(address(0x9541)); // Non-relayer address
+        // vm.expectRevert("not relayer");
```
Comment (mialbu) [inline test/FungibleToken.t.sol]: The newest version requires the relayer, so it should revert here now.

```suggestion
        vm.expectRevert("not relayer");
```

File: test/FungibleToken.t.sol [hunk 31]
```diff
@@ -0,0 +1,764 @@
...
+            tokenDepositRoot,
+            signatures,
+            depositData
+        );
+    }
+
+    function test_DepositTokenWithExceedMaxDeposits() public {
+        MockERC20(neoXTokenA).mint(address(bridgeImpl), 100 ether);
+        vm.prank(governor);
+        bridgeImpl.registerToken(neoXTokenA, validConfigA);
```
Comment (mialbu) [inline test/FungibleToken.t.sol:714]:
```suggestion
        // validConfigA.maxDeposits is 2
        bridgeImpl.registerToken(neoXTokenA, validConfigA);
```

File: test/FungibleToken.t.sol [hunk 32]
```diff
@@ -0,0 +1,764 @@
...
+            amount: 200,
+            nonce: 2
+        });
+        BridgeLib.DepositData memory d2 = BridgeLib.DepositData({
+            to: payable(governor),
+            amount: 200,
+            nonce: 3
+        });
+        depositData[0] = d0;
+        depositData[1] = d1;
```
Comment (mialbu) [inline test/FungibleToken.t.sol:733]: I'm a bit confused as to why the third depositData has not been added here. If the data is not needed, I assume none of the data is needed. Can't we just use an empty array of type `BridgeLib.DepositData` with size 3 for this test and get the same revert reason?

**Review by mialbu — CHANGES_REQUESTED** [review]
Great tests! 👏🙏

Most comments are about the format. Please have a look at them. If you're adding more test cases I'd appreciate it if you could do that in a different PR, so that it's easier for the reviewer. Took me quite to have a look at everything here. 😅

I very much liked the comments in the BridgeTest. It makes it a lot easier to read the test code if you can read in natural language what the test should do. Could you add some description to the TokenBridge tests as well? Especially with the more complex scenarios there. This would be much appreciated and will increase maintainability by a lot if we need to adapt tests for contract upgrades. Also, this will likely make it easier for auditors to understand the specific expected behavior more precisely.

So, in summary, please generally...
- ...use camel case for variables, e.g., instead of `getTokenbridge` use `getTokenBridge`.
- ...remove unnecessary blocks with multiple empty lines
- ...make sure every file has an empty line at the end of the file
- ...remove the `console.log`s (mainly in the token bridge test file)
- ...add some description to the token bridge tests (some inline comments and on top what the test wants to accomplish, e.g., `If the contract has insufficient funds, the deposit amounts should be made available for claiming.`

File: contracts/tests/MockERC20.sol (hunk 1 above)
Comment (moonpyt) [reply contracts/tests/MockERC20.sol]: Have been modified

File: contracts/tests/SigUtils.sol (hunk 2 above)
Comment (moonpyt) [reply contracts/tests/SigUtils.sol]: Have been modified

File: contracts/tests/SigUtils.sol (hunk 3 above)
Comment (moonpyt) [reply contracts/tests/SigUtils.sol:3]: Have been modified

File: contracts/tests/SigUtils.sol (hunk 4 above)
Comment (moonpyt) [reply contracts/tests/SigUtils.sol]: Have been modified

File: contracts/tests/TestBridge.sol (hunk 5 above)
Comment (moonpyt) [reply contracts/tests/TestBridge.sol:3]: Have been modified

File: .states/vm-cancun/state.json (hunk 6 above)
Comment (moonpyt) [reply .states/vm-cancun/state.json]: Have been modified

File: contracts/tests/SigUtils.sol (hunk 7 above)
Comment (moonpyt) [reply contracts/tests/SigUtils.sol]: Have been modified

File: contracts/tests/TestBridge.sol (hunk 8 above)
Comment (moonpyt) [reply contracts/tests/TestBridge.sol]: Have been modified

File: test/Bridge.t.sol (hunk 9 above)
Comment (moonpyt) [reply test/Bridge.t.sol]: Have been modified

File: test/Bridge.t.sol (hunk 10 above)
Comment (moonpyt) [reply test/Bridge.t.sol]: Have been modified

File: contracts/tests/TestBridge.sol
Comment (moonpyt) [reply contracts/tests/TestBridge.sol:1]: I added the case to use the getNeoN3Token method and changed getNeoN3Token for reading. Modified the getTokenConfig, getTokenbridgePaused methods to get the paused state of the bridge (because there is no way to get it, it can only be obtained by the variable value)

File: contracts/tests/TestBridge.sol (hunk 11 above)
Comment (moonpyt) [reply contracts/tests/TestBridge.sol]: Have been modified

File: test/Bridge.t.sol (hunk 12 above)
Comment (moonpyt) [reply test/Bridge.t.sol]: Have been modified

File: test/Bridge.t.sol (hunk 13 above)
Comment (moonpyt) [reply test/Bridge.t.sol]: Have been modified

File: test/Bridge.t.sol (hunk 14 above)
Comment (moonpyt) [reply test/Bridge.t.sol]: registerToken/unregisterToken pauseToken/unPausetoken depositToken/withdrawToken  all the successful case of these methods, have added a check event case

File: test/Bridge.t.sol (hunk 15 above)
Comment (moonpyt) [reply test/Bridge.t.sol]: Have been modified

File: test/Bridge.t.sol (hunk 17 above)
Comment (moonpyt) [reply test/Bridge.t.sol]: Have been modified

File: test/Bridge.t.sol (hunk 18 above)
Comment (moonpyt) [reply test/Bridge.t.sol]: Have been modified

File: test/BridgeManagement.t.sol (hunk 19 above)
Comment (moonpyt) [reply test/BridgeManagement.t.sol]: Have been modified

File: test/BridgeManagement.t.sol (hunk 21 above)
Comment (moonpyt) [reply test/BridgeManagement.t.sol]: Have been modified

File: test/FungibleToken.t.sol (hunk 22 above)
Comment (moonpyt) [reply test/FungibleToken.t.sol]: Have been modified

File: test/FungibleToken.t.sol (hunk 24 above)
Comment (moonpyt) [reply test/FungibleToken.t.sol]: Have been modified,Several test cases were added using any 6 nodes/any 5 nodes/all 7 nodes to depositToken and withdrawToken

File: test/FungibleToken.t.sol (hunk 27 above)
Comment (moonpyt) [reply test/FungibleToken.t.sol]: Have been modified

File: test/FungibleToken.t.sol (hunk 29 above)
Comment (moonpyt) [reply test/FungibleToken.t.sol]: Have been modified

File: test/FungibleToken.t.sol (hunk 30 above)
Comment (moonpyt) [reply test/FungibleToken.t.sol]: Have been modified

File: test/FungibleToken.t.sol (hunk 31 above)
Comment (moonpyt) [reply test/FungibleToken.t.sol:714]: Have been modified

File: test/FungibleToken.t.sol (hunk 32 above)
Comment (moonpyt) [reply test/FungibleToken.t.sol:733]: Have been modified

File: test/Bridge.t.sol (hunk 16 above)
Comment (moonpyt) [reply test/Bridge.t.sol:244]: Waiting for your modification

File: test/FungibleToken.t.sol (hunk 28 above)
Comment (moonpyt) [reply test/FungibleToken.t.sol]: Waiting for your modification

File: test/Bridge.t.sol (hunk 20 above)
Comment (moonpyt) [reply test/Bridge.t.sol]: The bridge.t.sol file needs this to initialize the management contract before it can execute other cases

File: test/FungibleToken.t.sol (hunk 23 above)
Comment (moonpyt) [reply test/FungibleToken.t.sol]: Modified use to funder

File: test/FungibleToken.t.sol (hunk 25 above)
Comment (moonpyt) [reply test/FungibleToken.t.sol:277]: Have been modified

File: test/FungibleToken.t.sol (hunk 23 above)
Comment (moonpyt) [reply test/FungibleToken.t.sol]: "addresses that don't have any role in the bridge management"  - Have been modified

File: test/FungibleToken.t.sol (hunk 26 above)
Comment (moonpyt) [reply test/FungibleToken.t.sol]: Have been modified，Some cases use different transfer amounts，It's not just 100 or 100ether

File: test/Bridge.t.sol (hunk 20 above)
Comment (mialbu) [reply test/Bridge.t.sol]: My mistake, GitHub shows at least 4 lines, so the lines before the one I marked are shown as well here, but I think I only meant the `uint256[] public validatorsKeys;`.

Comment (moonpyt) [reply test/Bridge.t.sol]: Ok, modified

File: test/Bridge.t.sol (hunk 16 above)
Comment (mialbu) [reply test/Bridge.t.sol:244]: I'm not doing that right now, you don't need to wait for this.

I've opened an issue for it (https://github.com/bane-labs/bridge-evm-contracts/issues/123).

File: test/FungibleToken.t.sol (hunk 28 above)
Comment (mialbu) [reply test/FungibleToken.t.sol]: Which modification do you mean, @moonpyt ?

File: test/Bridge.t.sol (hunk 16 above)
Comment (moonpyt) [reply test/Bridge.t.sol:244]: Well.

File: test/FungibleToken.t.sol (hunk 28 above)
Comment (moonpyt) [reply test/FungibleToken.t.sol]: My mistake,Checking withdrwal state is something I do，                                                                              
but I think depositToken method may need check tokenValue < config.minAmount and  tokenValue > config.maxAmount

Comment (mialbu) [reply test/FungibleToken.t.sol]: That's not necessary in the deposit functions. These values are already checked on the Neo N3 side when someone deposits gas or a token. These min/max values are used when someone wants to bridge to the other chain.

The same applies for withdrawals: The min/max values are checked here in this contract on Neo X, then the check is not done again when the relayer distributes the gas/tokens on Neo N3.

Comment (moonpyt) [reply test/FungibleToken.t.sol]: Ok, got it

**Review by mialbu — APPROVED** [review]
LGTM

**Comment by mialbu** [comment]
I'll merge this. Please open a new PR if there's a test you'd like to add in the future.
